packages/appium/lib/appium.js

```diff
@@ -276,34 +275,27 @@ class AppiumDriver extends BaseDriver {
   }
 
   attachUnexpectedShutdownHandler (driver, innerSessionId) {
-    const removeSessionFromMasterList = (cause = new Error('Unknown error')) => {
-      log.warn(`Closing session, cause was '${cause.message}'`);
+    const onShutdown = (cause = new Error('Unknown error')) => {
+      log.warn(`Ending session, cause was '${cause.message}'`);
+
+      if (this.sessionPlugins[innerSessionId]) {
+        for (const plugin of this.sessionPlugins[innerSessionId]) {
+          if (_.isFunction(plugin.onUnexpectedShutdown)) {
+            log.debug(`Plugin ${plugin.name} defines an unexpected shutdown handler; calling it now`);
```

Comment:
note that `plugin.name` can theoretically be an empty string. you might want `plugin.constructor.pluginName`?

Reply:
Is the order of plugins relevant?

Should `sessionPlugins` really be a `Set`?

Reply:
the order is relevant. they do a russian nesting doll sort of thing so depending on the order in which they are declared on the CLI, the ultimate behaviour might be different.

Reply:
and here plugin.name has been assigned by us so we should be able to trust it